sound/soc/amd/acp/acp-sdw-sof-mach.c

```diff
@@ -0,0 +1,1300 @@
+	if (!dai_links)
+		return -ENOMEM;
+
+	/* allocate codec conf, will be populated when dailinks are created */
+	codec_conf = devm_kcalloc(dev, codec_conf_num, sizeof(*codec_conf),
+				  GFP_KERNEL);
+	if (!codec_conf)
+		return -ENOMEM;
+
+	/* SDW */
+	if (!sdw_be_num)
+		goto DMIC;
+
+	for (; adr_link->num_adr; adr_link++) {
+		/*
+		 * If there are two or more different devices on the same sdw link, we have to
+		 * append the codec type to the dai link name to prevent duplicated dai link name.
+		 * The same type devices on the same sdw link will be in the same
+		 * snd_soc_acpi_adr_device array. They won't be described in different adr_links.
+		 */
+		for (i = 0; i < adr_link->num_adr; i++) {
+			/* find codec info to get dai_num */
+			codec_index = find_codec_info_part(adr_link->adr_d[i].adr);
+			if (codec_index < 0)
+				return codec_index;
+			if (codec_info_list[codec_index].dai_num > 1) {
+				append_dai_type = true;
+				goto out;
+			}
+			for (j = 0; j < i; j++) {
+				if ((SDW_PART_ID(adr_link->adr_d[i].adr) !=
+				    SDW_PART_ID(adr_link->adr_d[j].adr)) ||
+				    (SDW_MFG_ID(adr_link->adr_d[i].adr) !=
+				    SDW_MFG_ID(adr_link->adr_d[j].adr))) {
+					append_dai_type = true;
+					goto out;
+				}
+			}
+		}
+	}
+out:
+
+	/* generate DAI links by each sdw link */
+	for (adr_link = mach_params->links ; adr_link->num_adr; adr_link++) {
+		for (i = 0; i < adr_link->num_adr; i++) {
+			const struct snd_soc_acpi_endpoint *endpoint;
+
+			endpoint = adr_link->adr_d[i].endpoints;
+
+			/* this group has been generated */
+			if (endpoint->aggregated &&
+			    group_generated[endpoint->group_id])
+				continue;
+
+			/* find codec info to get dai_num */
+			codec_index = find_codec_info_part(adr_link->adr_d[i].adr);
+			if (codec_index < 0)
+				return codec_index;
+
+			for (j = 0; j < codec_info_list[codec_index].dai_num ; j++) {
+				ret = create_sdw_dailink(card, &link_index, dai_links,
+							 sdw_be_num,
+							 adr_link,
+							 codec_conf, codec_conf_num,
+							 &be_id, &codec_conf_index,
+							 &ignore_acp_dmic, append_dai_type, i, j);
+				if (ret < 0) {
+					dev_err(dev, "failed to create dai link %d\n", link_index);
+					return ret;
+				}
+			}
+
+			if (aggregation && endpoint->aggregated)
+				group_generated[endpoint->group_id] = true;
+		}
+	}
+
+DMIC:
+	/* dmic */
+	if (dmic_num > 0) {
+		if (ignore_acp_dmic) {
+			dev_warn(dev, "Ignoring ACP DMIC\n");
+			goto CARD_UPDATE;
+		}
+		be_id = SDW_DMIC_DAI_ID;
+		ret = init_simple_dai_link(dev, dai_links + link_index, &be_id, "acp-dmic-codec",
+					   0, 1, // DMIC only supports capture
+					   "acp-sof-dmic", "dmic-codec", "dmic-hifi",
+					   sof_sdw_dmic_init, NULL);
+		if (ret)
+			return ret;
+		link_index++;
+	}
+
+CARD_UPDATE:
+
+	card->dai_link = dai_links;
+	card->num_links = num_links;
+
+	card->codec_conf = codec_conf;
+	card->num_configs = codec_conf_num;
+
+	return 0;
+}
```

Comment:
I am not sure what the best solution is for this machine driver code...

On one hand this looks like a copy-paste of the Intel machine driver.
On the other hand I don't think anyone is ready for common code, we have too many platform-specific things and different timelines that would make sharing difficult.

Reply:
Yeah I would tend to agree the bits that are fairly vendor agnostic should be able to be exported, the functions pointed out above seem kinda reasonable. I would also suggest the code that parses the acpi_* structs and turns them into DAI links as well, but perhaps after my refactoring.

Reply:
@charleskeepax : will wait for your code changes.

Reply:
The bulk of my refactoring is merged into the sof repo, the only not merged patch you might want to look at is:

https://github.com/thesofproject/linux/pull/4848/commits/c0a9f4cfef2eaedf3f2390c36a89f79d147cab7e

But certainly enough for you have a look and decide how much of it you want to pull in here.

Reply:
@plbossart  & @charleskeepax : will start working on refactoring the code to have common codec helper functions